### PR TITLE
Dataflow builder dustoff

### PR DESCRIPTION
**Pull request by frankmcsherry**
### Motivation

   * This PR refactors existing code.

This PR reorganizes optimization in the coordinator, and `DataflowBuilder`. 

### Description

The coordinator no longer calls optimization with enabled indexes, as this is a potential mis-optimization when the set of available indexes changes before dataflow building. Instead, available indexes are used (continue to be used) only in `DataflowDescription` optimization, in `transform/dataflow.rs`. This prevents the coordinator from needing to reason about which indexes to present as available to optimization.

`DataflowBuilder` is re-organized to make building of sinks and indexes depend on descriptions, respectively `SinkDesc` and `IndexDesc`. This information was previously partially extracted from the bound catalog, and meant that items to be built must already be installed in the catalog (reported as frustrating). They may now be installed, but will not be consulted and the description is used instead.

For index building, the prevention of building for disabled indexes is moved outside the `build_index_dataflow` method. A helper method `prepare_index_build` is added to `Coordinator` (in `dataflow_builder.rs`) which extracts the name to use, index description, and whether the index is enabled. This information is the supplied directly to the index builder, who more clearly just builds the described index.

### Tips for reviewer

This has not be tested yet, and we'll see what CI says. Ideally it changes functionality not at all, but if e.g. enabled index information were leaking in to optimization, this now prevents that and may result in different plans.

### Checklist

- [ ] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

File: src/transform/src/lib.rs [hunk 1]
```diff
@@ -340,17 +340,22 @@ impl Optimizer {
...
+    ///
+    /// These optimizations are performed with no information about available arrangements,
+    /// which makes them suitable for pre-optimization before dataflow deployment.
     pub fn optimize(
         &mut self,
         mut relation: MirRelationExpr,
-        indexes: &HashMap<GlobalId, Vec<(GlobalId, Vec<MirScalarExpr>)>>,
     ) -> Result<expr::OptimizedMirRelationExpr, TransformError> {
-        self.transform(&mut relation, indexes)?;
+        self.transform(&mut relation, &HashMap::new())?;
```
Comment (mjibson) [inline src/transform/src/lib.rs:351]: It looks like the only leftover call to `transform` is from `dataflow.rs`, which includes indexes. Is that supposed to be the case?

**Review by mjibson — APPROVED** [review]
LGTM? I may revert some of the previous fixes to see if this would have caught them.

File: src/transform/src/lib.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/transform/src/lib.rs:351]: Yes, that is the one place we actually mean to invoke the method with specific indexes (as part of the last optimization before building the dataflow, as a function of the indexes that are imported).

**Comment by mjibson** [comment]
This looks totally unrelated to our two recent index-related bugs, right? Without a test change it's hard to tell what exactly this would prevent.

**Comment by frankmcsherry** [comment]
It prevents several things by way of API change. It prevents e.g. the optimizer team responding to the signals about available arrangements provided by the coordinator (by not providing them). It reduces (imo) confusion about what dataflow building does, by requiring more explicit arguments and with more specific behavior. It explicitly does not try to change any behavior.

**Comment by frankmcsherry** [comment]
Want to recheck on the LGTM @mjibson. The intent here is just a clean up, rather than adding new features or removing existing errors (potentially changing query plans, but it seems like not actually, which is good news). It is not "the end" of work on `DataflowBuilder`, but a bit of pre-work to get the methods doing imo clearer things.

**Comment by mjibson** [comment]
Yes, LGTM.

**Comment by frankmcsherry** [comment]
Thanks!

**Pull request merged**